src/clinical/service-worker-thread/tasks.ts

```diff
@@ -116,6 +118,20 @@ const sortRecordsByColumn = (
 	return valueSort;
 };
 
+// Sort Invalid Records to Top
+const sortInvalidRecords = (
+	currentRecord: ClinicalInfo,
+	nextRecord: ClinicalInfo,
+	errors: ClinicalErrorsResponseRecord[],
+) => {
+	const currentRecordIsInvalid = errors.some((record) => record.donorId === currentRecord.donor_id);
+	const nextRecordIsInvalid = errors.some((record) => record.donorId === nextRecord.donor_id);
+	const invalidSort =
+		currentRecordIsInvalid === nextRecordIsInvalid ? 0 : currentRecordIsInvalid ? -1 : 1;
+
+	return invalidSort;
+};
```

Comment:
This looks logically correct, but I can see how this is very slow. We need to iterate the errors list twice for every element in our array. If before iterating over your array you prepare a map of donorId to validity state and pass that as an argument to the sort function then you can just look up the validity of each donorId instead of checking the errors array.

```ts
const sortByValidity = (errors: ClinicalErrorsResponseRecord[]) => {
	const donorIdsWithErrors = errors.reduce<Set<any>>((acc, error) => acc.add(error.donorId), new Set());
	return (currentRecord: ClinicalInfo, nextRecord: ClinicalInfo): number => {
		const currentRecordIsInvalid = donorIdsWithErrors.has(currentRecord.donor_id);
		const nextRecordIsInvalid = donorIdsWithErrors.has(nextRecord.donor_id);
		return currentRecordIsInvalid === nextRecordIsInvalid ? 0 : currentRecordIsInvalid ? -1 : 1;
	};
};
```

Another way to improve this is to build 2 arrays, one of the valid donors and one of the invalid donors, sort each of those arrays, and then combine them. This way you don't need to use a compareFunction to sort valid vs invalid.

Reply:
OK I've updated this to use Sets and avoid multiple iterations, it's definitely a little faster now too
https://github.com/icgc-argo/argo-clinical/pull/1184/commits/4281c6a3d7b2f086d68369f3920e43badc8bbf53